SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp

```diff
@@ -11,46 +11,143 @@ namespace
+	}
+
+	if (!RuntimeIP.IsEmpty())
+	{
+		Command.Append(FString::Printf(TEXT(" --runtime_ip=%s"), *RuntimeIP));
+		UE_LOG(LogSpatialCommandUtils, Verbose, TEXT("Trying to start spatial service with exposed runtime ip: %s"), *RuntimeIP);
+	}
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
+	if (!bSuccess)
+	{
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial start service failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
 	}
-	return FPlatformProcess::CreateProc(*SpatialGDKServicesConstants::SpatialExe, *Arguments, bLaunchDetached, bLaunchHidden, bLaunchReallyHidden, OutProcessID, PriorityModifier, OptionalWorkingDirectory, PipeWriteChild, PipeReadChild);
+
+	return bSuccess;
 }
 
-bool SpatialCommandUtils::AttemptSpatialAuth(bool bIsRunningInChina)
+bool SpatialCommandUtils::StopSpatialService(bool bIsRunningInChina, const FString& DirectoryToRun, FString& OutResult, int32& OutExitCode)
 {
-	FString SpatialInfoResult;
-	FString StdErr;
-	int32 ExitCode;
-	ExecuteSpatialCommand(TEXT("auth login"), &ExitCode, &SpatialInfoResult, &StdErr, bIsRunningInChina);
+	FString Command = TEXT("service stop");
 
-	bool bSuccess = ExitCode == 0;
+	if (bIsRunningInChina)
+	{
+		Command += ChinaEnvironmentArgument;
+	}
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
+	if (!bSuccess)
+	{
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial stop service failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
+	}
+
+	return bSuccess;
+}
+
+bool SpatialCommandUtils::BuildWorkerConfig(bool bIsRunningInChina, const FString& DirectoryToRun, FString& OutResult, int32& OutExitCode)
+{
+	FString Command = TEXT("worker build build-config");
+
+	if (bIsRunningInChina)
+	{
+		Command += ChinaEnvironmentArgument;
+	}
+
+	FSpatialGDKServicesModule::ExecuteAndReadOutput(*SpatialGDKServicesConstants::SpatialExe, Command, DirectoryToRun, OutResult, OutExitCode);
+
+	bool bSuccess = OutExitCode == 0;
 	if (!bSuccess)
 	{
-		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial auth login failed. Error Code: %d, Error Message: %s"), ExitCode, *SpatialInfoResult);
+		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial build worker config failed. Error Code: %d, Error Message: %s"), OutExitCode, *OutResult);
 	}
 
 	return bSuccess;
 }
+
+FProcHandle SpatialCommandUtils::LocalWorkerReplace(const FString& ServicePort, const FString& OldWorker, const FString& NewWorker, bool bIsRunningInChina, bool bLaunchDetached, bool bLaunchHidden, bool bLaunchReallyHidden, uint32* OutProcessID, int32 PriorityModifier, const TCHAR* OptionalWorkingDirectory, void* PipeWriteChild, void* PipeReadChild)
+{
+	FString Command = TEXT("worker build build-config");
+
+	if (ServicePort.IsEmpty())
```

Comment:
Is there either a way to return something empty here?
Alternatively should these be checks instead? As currently it will detect it's wrong and carry on anyway.

Reply:
I was thinking about checks but I have been told asserts should only be placed where yo think it should be impossible to hit. This should error and return by that school of thought. If that is not what we go by then I can put in a check.